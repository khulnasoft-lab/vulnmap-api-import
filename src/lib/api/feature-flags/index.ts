import type { requestsManager } from 'snyk-request-manager';
import * as debugLib from 'debug';

const debug = debugLib('vulnmap:get-feature-flag');

export async function getFeatureFlag(
  requestManager: requestsManager,
  featureFlagName: string,
  orgId: string,
): Promise<boolean> {
  debug(
    `Checking if feature flag ${featureFlagName} is enabled for Vulnmap org ${orgId}`,
  );
  try {
    const url = `cli-config/feature-flags/${featureFlagName}?org=${orgId}`;
    const res = await requestManager.request({
      verb: 'get',
      url: url,
      useRESTApi: false,
    });

    debug(`Feature flag ${featureFlagName} is enabled for Org ${orgId}`);

    const enabled: boolean = res.data['ok'];

    return enabled;
  } catch (err: any) {
    const res = err.message?.response?.data;
    const message = res?.userMessage || res?.message || err.message?.message;

    if (
      res &&
      res.ok === false &&
      res?.userMessage?.includes('feature enabled')
    ) {
      debug(`Feature flag ${featureFlagName} is not enabled for Org ${orgId}`);
      return false;
    }

    debug(
      `Could not fetch the ${featureFlagName} feature flag for ${orgId}. Error: ${JSON.stringify(
        err,
      )}`,
    );
    throw new Error(
      `Could not fetch the ${featureFlagName} feature flag for ${orgId}. ${message}`,
    );
  }
}
