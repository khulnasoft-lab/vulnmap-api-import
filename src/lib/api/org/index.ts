import 'source-map-support/register';
import type { requestsManager } from 'snyk-request-manager';
import * as debugLib from 'debug';
import * as qs from 'querystring';
import { getApiToken } from '../../get-api-token';
import { getVulnmapHost } from '../../get-vulnmap-host';
import type {
  VulnmapProject,
  RESTTargetResponse,
  RESTProjectData,
  VulnmapTarget,
} from '../../types';

const debug = debugLib('vulnmap:api-group');

export interface IntegrationsListResponse {
  [name: string]: string;
}

export async function listIntegrations(
  requestManager: requestsManager,
  orgId: string,
): Promise<IntegrationsListResponse> {
  getApiToken();
  getVulnmapHost();
  debug('Listing integrations for org: ' + orgId);

  if (!orgId) {
    throw new Error(
      `Missing required parameters. Please ensure you have set: orgId.
      \nFor more information see: https://vulnmap.docs.apiary.io/#reference/integrations/list`,
    );
  }

  const res = await requestManager.request({
    verb: 'get',
    url: `/org/${orgId.trim()}/integrations`,
    body: JSON.stringify({}),
  });

  const statusCode = res.statusCode || res.status;
  if (!statusCode || statusCode !== 200) {
    throw new Error(
      'Expected a 200 response, instead received: ' +
        JSON.stringify({ data: res.data || res.body, status: statusCode }),
    );
  }
  return res.data || {};
}

const defaultDisabledSettings = {
  'new-issues-remediations': {
    enabled: false,
    issueType: 'none',
    issueSeverity: 'high',
  },
  'project-imported': {
    enabled: false,
  },
  'test-limit': {
    enabled: false,
  },
  'weekly-report': {
    enabled: false,
  },
};

interface NotificationSettings {
  [name: string]: {
    enabled?: boolean;
    issueSeverity?: string;
    issueType?: string;
  };
}

export async function setNotificationPreferences(
  requestManager: requestsManager,
  orgId: string,
  orgName: string,
  settings: NotificationSettings = defaultDisabledSettings,
): Promise<IntegrationsListResponse> {
  getApiToken();
  getVulnmapHost();
  debug(`Disabling notifications for org: ${orgName} (${orgId})`);

  if (!orgId) {
    throw new Error(
      `Missing required parameters. Please ensure you have set: orgId, settings.
      \nFor more information see: https://vulnmap.docs.apiary.io/#reference/organizations/the-vulnmap-organization-for-a-request/set-notification-settings`,
    );
  }
  try {
    const res = await requestManager.request({
      verb: 'put',
      url: `/org/${orgId.trim()}/notification-settings`,
      body: JSON.stringify(settings),
    });

    const statusCode = res.statusCode || res.status;
    if (!statusCode || statusCode !== 200) {
      throw new Error(
        'Expected a 200 response, instead received: ' +
          JSON.stringify({ data: res.data, status: statusCode }),
      );
    }
    return res.data || {};
  } catch (e) {
    debug('Failed to update notification settings for ', orgId, e);
    throw e;
  }
}

export async function deleteOrg(
  requestManager: requestsManager,
  orgId: string,
): Promise<unknown> {
  getApiToken();
  getVulnmapHost();
  debug(`Deleting org: "${orgId}"`);

  if (!orgId) {
    throw new Error(
      `Missing required parameters. Please ensure you have set: orgId.
      \nFor more information see: https://vulnmap.docs.apiary.io/#reference/organizations/manage-organization/remove-organization`,
    );
  }
  const res = await requestManager.request({
    verb: 'delete',
    url: `/org/${orgId}`,
    body: JSON.stringify({}),
  });
  const statusCode = res.statusCode || res.status;
  if (!statusCode || statusCode !== 204) {
    throw new Error(
      'Expected a 204 response, instead received: ' +
        JSON.stringify({ data: res.data, status: statusCode }),
    );
  }
  return res.data;
}

export interface ProjectsResponse {
  org: {
    id: string;
  };
  projects: VulnmapProject[];
}

interface RESTProjectsResponse {
  data: RESTProjectData[];
  jsonapi: {
    version: string;
  };
  links: {
    first: string;
    last: string;
    next: string;
    prev: string;
    related: string;
    self: string;
  };
}

interface ProjectsFilters {
  name?: string; // If supplied, only projects that have a name that starts with this value will be returned
  origin?: string; //If supplied, only projects that exactly match this origin will be returned
  type?: string; //If supplied, only projects that exactly match this type will be returned
  isMonitored?: boolean; // If set to true, only include projects which are monitored, if set to false, only include projects which are not monitored
  targetId?: string; // The target public ID
  limit?: number; // how many results per page, defaults to 10
}

export async function listProjects(
  requestManager: requestsManager,
  orgId: string,
  filters?: ProjectsFilters,
): Promise<ProjectsResponse> {
  getApiToken();
  getVulnmapHost();
  debug(
    `Listing all projects for org: ${orgId} with filter ${JSON.stringify(
      filters,
    )}`,
  );
  if (!orgId) {
    throw new Error(
      `Missing required parameters. Please ensure you have set: orgId, settings.
        \nFor more information see: https://vulnmap.docs.apiary.io/#reference/projects/all-projects/list-all-projects`,
    );
  }

  const projects = await listAllProjects(requestManager, orgId, filters);

  return {
    org: {
      id: orgId,
    },
    projects: projects,
  };
}

async function listAllProjects(
  requestManager: requestsManager,
  orgId: string,
  filters?: ProjectsFilters,
): Promise<VulnmapProject[]> {
  let lastPage = false;
  const projectsList: VulnmapProject[] = [];
  let pageCount = 1;
  let nextPageLink: string | undefined = undefined;
  while (!lastPage) {
    try {
      debug(
        `Fetching page ${pageCount} to list all projects for org ${orgId}\n`,
      );
      const {
        projects,
        next,
      }: {
        projects: VulnmapProject[];
        next?: string;
      } = await getProjectsPage(requestManager, orgId, filters, nextPageLink);

      projectsList.push(...projects);
      next
        ? ((lastPage = false), (nextPageLink = next))
        : ((lastPage = true), (nextPageLink = ''));
      pageCount++;
    } catch (e) {
      debug('Failed to get projects for ', orgId, e);
      throw e;
    }
  }
  return projectsList;
}

async function getProjectsPage(
  requestManager: requestsManager,
  orgId: string,
  filters?: ProjectsFilters,
  nextPageLink?: string,
): Promise<{ projects: VulnmapProject[]; next?: string }> {
  const query = qs.stringify({
    version: '2022-09-15~beta',
    ...filters,
  });

  const url = nextPageLink ?? `/orgs/${orgId.trim()}/projects?${query}`;

  const res = await requestManager.request({
    verb: 'get',
    url: url,
    useRESTApi: true,
  });

  const statusCode = res.statusCode || res.status;
  if (!statusCode || statusCode !== 200) {
    throw new Error(
      'Expected a 200 response, instead received: ' +
        JSON.stringify({ data: res.data, status: statusCode }),
    );
  }

  const response = res.data as RESTProjectsResponse;

  const projects = convertToVulnmapProject(response.data);
  const next = response.links.next;

  return { projects, next };
}

function convertToVulnmapProject(
  projectData: RESTProjectData[],
): VulnmapProject[] {
  const projects: VulnmapProject[] = [];

  for (const project of projectData) {
    const projectTmp: VulnmapProject = {
      id: project.id,
      branch: project.attributes.targetReference,
      created: project.attributes.created,
      origin: project.attributes.origin,
      name: project.attributes.name,
      type: project.attributes.type,
      status: project.attributes.status,
    };
    projects.push(projectTmp);
  }

  return projects;
}
export interface TargetFilters {
  remoteUrl?: string;
  limit?: number;
  isPrivate?: boolean;
  origin?: string;
  displayName?: string;
  excludeEmpty?: boolean;
}
export async function listTargets(
  requestManager: requestsManager,
  orgId: string,
  config?: TargetFilters,
): Promise<{ targets: VulnmapTarget[] }> {
  getApiToken();
  getVulnmapHost();
  debug(`Listing all targets for org: ${orgId}`);

  if (!orgId && orgId.length > 0) {
    throw new Error(
      `Missing required parameters. Please ensure you have set: orgId, settings.
        \nFor more information see: https://apidocs.vulnmap.khulnasoft.com/?version=2022-09-15%7Ebeta#get-/orgs/-org_id-/targets`,
    );
  }

  const targets = await listAllVulnmapTargets(requestManager, orgId, config);

  return { targets };
}

export async function listAllVulnmapTargets(
  requestManager: requestsManager,
  orgId: string,
  config?: TargetFilters,
): Promise<VulnmapTarget[]> {
  let lastPage = false;
  const targetsList: VulnmapTarget[] = [];
  let pageCount = 1;
  let nextPageLink: string | undefined = undefined;
  while (!lastPage) {
    try {
      debug(`Fetching page ${pageCount} of targets for orgId: ${orgId}\n`);
      const { targets, next }: { targets: VulnmapTarget[]; next?: string } =
        await getVulnmapTarget(requestManager, orgId, nextPageLink, config);

      targetsList.push(...targets);
      next
        ? ((lastPage = false), (nextPageLink = next))
        : ((lastPage = true), (nextPageLink = undefined));
      pageCount++;
    } catch (e) {
      debug('Failed to get targets for ', orgId, e);
      throw e;
    }
  }
  return targetsList;
}

export async function getVulnmapTarget(
  requestManager: requestsManager,
  orgId: string,
  nextPageLink?: string,
  config: {
    limit?: number;
    excludeEmpty?: boolean;
    origin?: string;
  } = {
    limit: 20,
    excludeEmpty: true,
  },
): Promise<{ targets: VulnmapTarget[]; next?: string }> {
  const query = qs.stringify({
    version: '2022-09-15~beta',
    ...config,
  });
  const url = nextPageLink ?? `/orgs/${orgId.trim()}/targets?${query}`;

  const res = await requestManager.request({
    verb: 'get',
    url: url,
    body: undefined,
    useRESTApi: true,
  });

  const statusCode = res.statusCode || res.status;
  if (!statusCode || statusCode !== 200) {
    throw new Error(
      'Expected a 200 response, instead received: ' +
        JSON.stringify({ data: res.data, status: statusCode }),
    );
  }

  const responseData = res.data as RESTTargetResponse;
  const targets = responseData.data;
  const { next } = responseData.links;

  return { targets, next };
}
