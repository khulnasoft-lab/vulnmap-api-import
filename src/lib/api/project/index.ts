import 'source-map-support/register';
import * as needle from 'needle';
import * as debugLib from 'debug';
import { getApiToken } from '../../get-api-token';
import { getVulnmapHost } from '../../get-vulnmap-host';
import type { requestsManager } from 'snyk-request-manager';
import type { VulnmapProject } from '../../types';
const debug = debugLib('vulnmap:api-project');

interface BulkProjectUpdateResponse {
  publicId: string;
  name: string;
}
export async function deleteProjects(
  orgId: string,
  projects: string[],
): Promise<{
  success: BulkProjectUpdateResponse[];
  failure: BulkProjectUpdateResponse[];
}> {
  const apiToken = getApiToken();
  if (!(orgId && projects)) {
    throw new Error(
      `Missing required parameters. Please ensure you have provided: orgId & projectIds.`,
    );
  }
  debug('Deleting projectIds:', projects.join(', '));

  try {
    const VULNMAP_API = getVulnmapHost();
    const body = {
      projects,
    };
    const res = await needle(
      'post',
      `${VULNMAP_API}/org/${orgId.trim()}/projects/bulk-delete`,
      body,
      {
        json: true,
        // eslint-disable-next-line @typescript-eslint/camelcase
        read_timeout: 30000,
        headers: {
          'content-type': 'application/json',
          Authorization: `token ${apiToken}`,
        },
      },
    );
    const statusCode = res.statusCode;
    if (!statusCode || statusCode !== 200) {
      throw new Error(
        `Expected a 200 response, instead received statusCode: ${
          res.statusCode
        },
          body: ${JSON.stringify(res.body)}`,
      );
    }
    return res.body;
  } catch (error) {
    debug('Could not delete project:', error.message || error);
    const err: {
      message?: string | undefined;
      innerError?: string;
    } = new Error('Could not delete project');
    err.innerError = error;
    throw err;
  }
}

export async function deactivateProject(
  requestManager: requestsManager,
  orgId: string,
  projectPublicId: string,
): Promise<boolean> {
  getApiToken();
  getVulnmapHost();
  if (!(orgId && projectPublicId)) {
    throw new Error(
      `Missing required parameters. Please ensure you have provided: orgId & projectId.`,
    );
  }
  debug(`De-activating project: ${projectPublicId}`);
  const url = `/org/${orgId.trim()}/project/${projectPublicId}/deactivate`;

  const res = await requestManager.request({
    verb: 'post',
    url: url,
    body: JSON.stringify({}),
    useRESTApi: false,
  });

  const statusCode = res.statusCode || res.status;
  if (!statusCode || statusCode !== 200) {
    debug(`Failed de-activating project projectId`);
    throw new Error(
      'Expected a 200 response, instead received: ' +
        JSON.stringify({ data: res.data, status: statusCode }),
    );
  }

  debug('Updated projects batch');
  return true;
}

export async function updateProject(
  requestManager: requestsManager,
  orgId: string,
  projectId: string,
  config: { branch: string },
): Promise<VulnmapProject> {
  getApiToken();
  getVulnmapHost();
  if (!orgId || !projectId) {
    throw new Error(
      `Missing required parameters. Please ensure you have set: orgId and projectId.
      \nFor more information see: https://vulnmap.docs.apiary.io/reference/projects/individual-project/update-a-project`,
    );
  }

  const body = {
    branch: config.branch,
  };

  const url = `/org/${orgId.trim()}/project/${projectId.trim()}`;
  const res = await requestManager.request({
    verb: 'put',
    url: url,
    body: JSON.stringify(body),
    useRESTApi: false,
  });

  const statusCode = res.statusCode || res.status;
  if (!statusCode || statusCode !== 200) {
    debug('Failed updating project: ' + projectId);
    throw new Error(
      'Expected a 200 response, instead received: ' +
        JSON.stringify({ data: res.data, status: statusCode }),
    );
  }

  const updatedProject: VulnmapProject = res.data;
  debug('Updated project: ' + projectId);
  return updatedProject;
}
