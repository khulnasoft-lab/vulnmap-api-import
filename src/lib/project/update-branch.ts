import debug = require('debug');
import type { requestsManager } from 'snyk-request-manager';
import { updateProject } from '../api/project';

export async function updateBranch(
  requestManager: requestsManager,
  project: {
    branch: string;
    projectPublicId: string;
  },
  defaultBranch: string,
  orgId: string,
  dryRun = false,
): Promise<{ updated: boolean }> {
  const { branch, projectPublicId } = project;
  let updated = false;
  try {
    if (branch != defaultBranch) {
      debug(
        `Default branch has changed for Vulnmap project ${projectPublicId}`,
      );
      if (!dryRun) {
        await updateProject(requestManager, orgId, project.projectPublicId, {
          branch: defaultBranch,
        });
      } else {
        debug(`Skipping API call to change branch in --dryRun mode`);
      }
      updated = true;
    } else {
      debug(`Default branch has not changed`);
    }
    return { updated };
  } catch (e) {
    throw new Error(
      `Failed to update project ${projectPublicId} via Vulnmap API. ERROR: ${e.message}`,
    );
  }
}
