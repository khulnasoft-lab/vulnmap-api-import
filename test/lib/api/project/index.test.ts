import { requestsManager } from 'snyk-request-manager';
import { updateProject } from '../../../../src/lib/api/project';
describe('UpdateProject', () => {
  const OLD_ENV = process.env;
  const requestManager = new requestsManager({
    userAgentPrefix: 'vulnmap-api-import:tests',
  });
  beforeAll(() => {
    process.env.VULNMAP_API = process.env.VULNMAP_API_TEST;
    process.env.VULNMAP_TOKEN = process.env.VULNMAP_TOKEN_TEST;
  });
  afterAll(() => {
    jest.restoreAllMocks();
    process.env = { ...OLD_ENV };
  }, 1000);

  it('Update project branch', async () => {
    jest.spyOn(requestManager, 'request').mockResolvedValue({
      data: {
        name: 'test',
        id: 'af137b96-6966-46c1-826b-2e79ac49bbd9',
        created: '2018-10-29T09:50:54.014Z',
        origin: 'github',
        type: 'maven',
        readOnly: false,
        testFrequency: 'daily',
        totalDependencies: 42,
        issueCountsBySeverity: {},
        imageId: 'sha256:caf27325b298a6730837023a8a342699c8b7b388b8d',
        imageTag: 'latest',
        imageBaseImage: 'alpine:3',
        imagePlatform: 'linux/arm64',
        imageCluster: 'Production',
        hostname: null,
        remoteRepoUrl: 'https://github.com/vulnmap/test.git',
        lastTestedDate: '2019-02-05T08:54:07.704Z',
        browseUrl:
          'https://app.vulnmap.khulnasoft.com/org/4a18d42f-0706-4ad0-b127-24078731fbed/project/af137b96-6966-46c1-826b-2e79ac49bbd9',
        importingUser: {},
        isMonitored: false,
        branch: 'newDefaultBranch',
        targetReference: null,
        tags: [],
        attributes: {},
        remediation: {},
      },
      status: 200,
    });

    const res = await updateProject(
      requestManager,
      'af137b96-6966-46c1-826b-2e79ac49bbxx',
      'af137b96-6966-46c1-826b-2e79ac49bbd9',
      { branch: 'newDefaultBranch' },
    );
    expect(res.branch).toMatchSnapshot();
  });

  it('Error if the requests fails', async () => {
    jest
      .spyOn(requestManager, 'request')
      .mockResolvedValue({ statusCode: 500, data: {} });
    expect(async () => {
      await updateProject(
        requestManager,
        'af137b96-6966-46c1-826b-2e79ac49bbxx',
        'af137b96-6966-46c1-826b-2e79ac49bbd9',
        { branch: 'newDefaultBranch' },
      );
    }).rejects.toThrowError(
      'Expected a 200 response, instead received: {"data":{},"status":500}',
    );
  }, 5000);
});
