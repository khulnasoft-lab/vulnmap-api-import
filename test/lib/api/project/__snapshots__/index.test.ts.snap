// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateProject Update project branch 1`] = `"newDefaultBranch"`;
